rust/oak/src/lib.rs

```diff
@@ -292,7 +292,6 @@ pub extern "C" fn oak_handle_grpc_call() {
         }
         None => {
             writeln!(logging_channel(), "gRPC call with no loaded Node").unwrap();
-            panic!("gRPC call with no loaded Node");
```

Comment:
This is not sufficient; we still panic all over the place within these functions. I think the proper solution would be to use [`catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) at the top level from any external entry point, and handle failures by doing something that is not undefined behaviour. In practice, I think panics do translate into `unreachable` or some other trap (https://webassembly.github.io/spec/core/syntax/instructions.html#syntax-instr-control) in wasm, but perhaps we should make that explicit. We could either emulate that behaviour, or perhaps add another exported function so that a module that is panicking can invoke that function and report the panic to the runtime (this would be in addition to the logging hook:

https://github.com/project-oak/oak/blob/53dd1b4d648fc75d8a0764b0c655cd7c7a0561c4/rust/oak/src/lib.rs#L263-L280

Reply:
The other option is that we do assume that panic always results in a trap when compiled to wasm (https://rustwasm.github.io/book/reference/code-size.html?highlight=panic#use-the-wasm-snip-tool), and then we can just have another layer of functions if we do want to make assertions about panics in tests, e.g. we may add an `oak_handle_grpc_call_inner` function that is not `extern` and we use that for testing, and then we let `oak_handle_grpc_call` just call that immediately (and nothing else). WDYT?

Reply:
(Ah, I misunderstood the scope of the problem.)

I think I prefer the first suggestion; assuming I'm understanding correctly, how about:
 - Putting `catch_unwind` into the standard implementations of all external entrypoints (`oak_initialize`, `oak_handle_grpc_call`) 
-  Changing all of their signatures to return an `i32` status, with a particular value for "caught a panic" that the `catch_unwind` code could emit.

The tests could then just check for that failure return, and we wouldn't need any test-specific wrappers/inner functions.

Reply:
That would prevent those functions to ever returning anything themselves though. Also the proper way for panic to manifest themselves would be traps in wasm, so I think we should make sure that's what's happening, one way or another.

Reply:
I'm not sure I agree that `panic`s should map to Wasm traps &ndash; if there's a higher-level alternative to catch the `panic` and terminate in a more orderly fashion, I think we should use it.

Reply:
Then what do we do about traps? We cannot prevent arbitrary Wasm code to trap, so we need to have a plan for that regardless (which may well just be "ignore them and terminate the node").

Reply:
I think that a Wasm trap will terminate the Node and pop out of `RunExportByName("oak_main")` with an error `ExecResult` value.

Reply:
I meant, what should we do from an observability point of view? Should we expose this to developers? Or somehow let this informative escape from the Oak runtime? (I don't think so)